lib/spack/spack/installer.py

```diff
@@ -0,0 +1,1554 @@
+    """
+    Convert seconds to hours, minutes, seconds
+
+    Args:
+        seconds (int): time to be converted in seconds
+
+    Return:
+        (str) String representation of the time as #h #m #.##fs
+    """
+    m, s = divmod(seconds, 60)
+    h, m = divmod(m, 60)
+
+    parts = []
+    if h:
+        parts.append("%dh" % h)
+    if m:
+        parts.append("%dm" % m)
+    if s:
+        parts.append("%.2fs" % s)
+    return ' '.join(parts)
+
+
+def _installed_from_cache(pkg, cache_only, explicit):
+    """
+    Install the package from binary cache
+
+    Args:
+        pkg (spack.package.PackageBase): the package to install from the binary
+            cache
+        cache_only (bool): only install from binary cache
+        explicit (bool): ``True`` if installing the package was explicitly
+            requested by the user, otherwise, ``False``
+
+    Return:
+        (bool) ``True`` if the package was installed from binary cache,
+            ``False`` otherwise
+    """
+    if _try_install_from_binary_cache(pkg, explicit):
+        tty.debug('Successfully installed {0} from binary cache'
+                  .format(pkg.unique_id))
+        _print_installed_pkg(pkg.spec.prefix)
+        spack.hooks.post_install(pkg.spec)
+        return True
+    elif cache_only:
+        tty.die('No binary for {0} found when cache-only specified'
+                .format(pkg.unique_id))
+
+    tty.debug('No binary for {0} found: installing from source'
+              .format(pkg.unique_id))
+    return False
+
+
+def _print_installed_pkg(message):
+    """
+    Output a message with a package icon.
+
+    Args:
+        message (str): message to be output
+    """
+    cwrite('@*g{[+]} ')
+    print(message)
+
+
+def _process_external_package(pkg, explicit):
+    """
+    Helper function to run post install hooks and register external packages.
+
+    Args:
+        pkg (Package): the external package
+        explicit (bool): if the package was requested explicitly by the user,
+            ``False`` if it was pulled in as a dependency of an explicit
+            package.
+    """
+    # TODO: Should this error out?
+    if not pkg.spec.external:
+        return
+
+    pre = '{s.name}@{s.version} :'.format(s=pkg.spec)
+    spec = pkg.spec
+
+    if spec.external_module:
+        tty.msg('{0} has external module in {1}'
+                .format(pre, spec.external_module))
+        tty.msg('{0} is actually installed in {1}'
+                .format(pre, spec.external_path))
+    else:
+        tty.msg("{0} externally installed in {1}"
+                .format(pre, spec.external_path))
+
+    try:
+        # Check if the package was already registered in the DB.
+        # If this is the case, then just exit.
+        rec = spack.store.db.get_record(spec)
+        tty.msg('{0} already registered in DB'.format(pre))
+
+        # Update the value of rec.explicit if it is necessary
+        _update_explicit_entry_in_db(pkg, rec, explicit)
```

Comment:
Is there a reason why functions that are otherwise left unchanged, like this one, have been extracted from `package.py`? Asking because this makes in general more difficult to understand if there have been changes due to the distributed builds or not. For instance in this case it seems to me that the function is exactly the same as it was, but I had to go to `package.py` and manually check line by line with the definition below.

Reply:
Apologies.  I did it for encapsulation as I responded previously.

I will try to remember to make a note of these circumstances in the future.